File: src/main/java/kirjanpito/ui/javafx/EntryRowModel.java
package kirjanpito.ui.javafx;

import javafx.beans.property.*;
import kirjanpito.db.Account;
import kirjanpito.db.Entry;

import java.math.BigDecimal;
import java.text.DecimalFormat;

/**
 * JavaFX-yhteensopiva Entry-rivi TableView:lle.
 * Käyttää JavaFX propertyja editoinnin mahdollistamiseksi.
 */
public class EntryRowModel {
    
    private final IntegerProperty rowNumber = new SimpleIntegerProperty();
    private final ObjectProperty<Account> account = new SimpleObjectProperty<>();
    private final StringProperty description = new SimpleStringProperty();
    private final ObjectProperty<BigDecimal> debit = new SimpleObjectProperty<>();
    private final ObjectProperty<BigDecimal> credit = new SimpleObjectProperty<>();
    private final StringProperty vatCode = new SimpleStringProperty();
    
    // Original entry for saving
    private Entry originalEntry;
    private boolean modified = false;
    
    private static final DecimalFormat CURRENCY_FORMAT = new DecimalFormat("#,##0.00");
    
    public EntryRowModel() {
        this.originalEntry = new Entry();
    }
    
    public EntryRowModel(int rowNum, Entry entry, Account account) {
        this.rowNumber.set(rowNum);
        this.originalEntry = entry;
        this.account.set(account);
        this.description.set(entry != null ? entry.getDescription() : "");
        
        if (entry != null) {
            if (entry.isDebit()) {
                this.debit.set(entry.getAmount());
                this.credit.set(null);
            } else {
                this.debit.set(null);
                this.credit.set(entry.getAmount());
            }
        }
        
        // Aseta ALV tilin perusteella
        updateVatFromAccount();
    }
    
    // Row number
    public int getRowNumber() { return rowNumber.get(); }
    public void setRowNumber(int value) { rowNumber.set(value); }
    public IntegerProperty rowNumberProperty() { return rowNumber; }
    
    // Account
    public Account getAccount() { return account.get(); }
    public void setAccount(Account value) { 
        account.set(value);
        if (originalEntry != null && value != null) {
            originalEntry.setAccountId(value.getId());
        }
        updateVatFromAccount();
        modified = true;
    }
    public ObjectProperty<Account> accountProperty() { return account; }
    
    public String getAccountDisplay() {
        Account acc = account.get();
        if (acc == null) return "";
        return acc.getNumber() + " " + acc.getName();
    }
    
    // Description
    public String getDescription() { return description.get(); }
    public void setDescription(String value) { 
        description.set(value);
        if (originalEntry != null) {
            originalEntry.setDescription(value);
        }
        modified = true;
    }
    public StringProperty descriptionProperty() { return description; }
    
    // Debit
    public BigDecimal getDebit() { return debit.get(); }
    public void setDebit(BigDecimal value) { 
        debit.set(value);
        if (value != null && value.compareTo(BigDecimal.ZERO) != 0) {
            credit.set(null); // Clear credit if debit is set
            if (originalEntry != null) {
                originalEntry.setDebit(true);
                originalEntry.setAmount(value);
            }
        }
        modified = true;
    }
    public ObjectProperty<BigDecimal> debitProperty() { return debit; }
    
    public String getDebitFormatted() {
        BigDecimal d = debit.get();
        return d != null && d.compareTo(BigDecimal.ZERO) != 0 ? CURRENCY_FORMAT.format(d) : "";
    }
    
    // Credit
    public BigDecimal getCredit() { return credit.get(); }
    public void setCredit(BigDecimal value) { 
        credit.set(value);
        if (value != null && value.compareTo(BigDecimal.ZERO) != 0) {
            debit.set(null); // Clear debit if credit is set
            if (originalEntry != null) {
                originalEntry.setDebit(false);
                originalEntry.setAmount(value);
            }
        }
        modified = true;
    }
    public ObjectProperty<BigDecimal> creditProperty() { return credit; }
    
    public String getCreditFormatted() {
        BigDecimal c = credit.get();
        return c != null && c.compareTo(BigDecimal.ZERO) != 0 ? CURRENCY_FORMAT.format(c) : "";
    }
    
    // VAT
    public String getVatCode() { return vatCode.get(); }
    public void setVatCode(String value) { vatCode.set(value); }
    public StringProperty vatCodeProperty() { return vatCode; }
    
    public void updateVatFromAccount() {
        Account acc = account.get();
        if (acc != null && acc.getVatRate() != null) {
            java.math.BigDecimal rate = acc.getVatRate();
            if (rate.compareTo(java.math.BigDecimal.ZERO) > 0) {
                vatCode.set(rate.stripTrailingZeros().toPlainString() + "%");
            } else {
                vatCode.set("");
            }
        } else {
            vatCode.set("");
        }
    }
    
    // Entry access
    public Entry getEntry() { return originalEntry; }
    public void setEntry(Entry entry) { this.originalEntry = entry; }
    
    public boolean isModified() { return modified; }
    public void setModified(boolean value) { this.modified = value; }
    
    public boolean isEmpty() {
        return account.get() == null && 
               (description.get() == null || description.get().isEmpty()) &&
               debit.get() == null && 
               credit.get() == null;
    }
    
    /**
     * Päivittää Entry-olion tiedot tämän rivin tiedoilla.
     */
    public void updateEntry() {
        if (originalEntry == null) {
            originalEntry = new Entry();
        }
        
        originalEntry.setRowNumber(rowNumber.get());
        originalEntry.setDescription(description.get());
        
        Account acc = account.get();
        if (acc != null) {
            originalEntry.setAccountId(acc.getId());
        }
        
        BigDecimal d = debit.get();
        BigDecimal c = credit.get();
        
        if (d != null && d.compareTo(BigDecimal.ZERO) != 0) {
            originalEntry.setDebit(true);
            originalEntry.setAmount(d);
        } else if (c != null && c.compareTo(BigDecimal.ZERO) != 0) {
            originalEntry.setDebit(false);
            originalEntry.setAmount(c);
        } else {
            originalEntry.setAmount(BigDecimal.ZERO);
        }
    }
}

File: src/main/kotlin/kirjanpito/ui/javafx/EntryRowModel.kt
package kirjanpito.ui.javafx

import javafx.beans.property.*
import kirjanpito.db.Account
import kirjanpito.db.Entry
import java.math.BigDecimal
import java.text.DecimalFormat

/**
 * JavaFX-yhteensopiva Entry-rivi TableView:lle.
 * Käyttää JavaFX propertyja editoinnin mahdollistamiseksi.
 */
class EntryRowModel(
    rowNum: Int = 0,
    entry: Entry? = null,
    account: Account? = null
) {
    private val rowNumber = SimpleIntegerProperty(rowNum)
    private val accountProperty = SimpleObjectProperty<Account>(account)
    private val description = SimpleStringProperty(entry?.getDescription() ?: "")
    private val debit = SimpleObjectProperty<BigDecimal>()
    private val credit = SimpleObjectProperty<BigDecimal>()
    private val vatCode = SimpleStringProperty("")
    
    // Original entry for saving
    var originalEntry: Entry? = entry ?: Entry()
        private set
    private var modified = false
    
    companion object {
        private val CURRENCY_FORMAT = DecimalFormat("#,##0.00")
    }
    
    init {
        entry?.let {
            if (it.isDebit()) {
                debit.set(it.getAmount())
                credit.set(null)
            } else {
                debit.set(null)
                credit.set(it.getAmount())
            }
        }
        
        // Aseta ALV tilin perusteella
        updateVatFromAccount()
    }
    
    // Row number
    fun getRowNumber() = rowNumber.get()
    fun setRowNumber(value: Int) = rowNumber.set(value)
    fun rowNumberProperty() = rowNumber
    
    // Account
    fun getAccount() = accountProperty.get()
    fun setAccount(value: Account?) {
        accountProperty.set(value)
        originalEntry?.let {
            value?.let { acc ->
                it.setAccountId(acc.id)
            }
        }
        updateVatFromAccount()
        modified = true
    }
    fun accountProperty() = accountProperty
    
    fun getAccountDisplay(): String {
        val acc = accountProperty.get()
        return acc?.let { "${it.number} ${it.name}" } ?: ""
    }
    
    // Description
    fun getDescription() = description.get()
    fun setDescription(value: String) {
        description.set(value)
        originalEntry?.setDescription(value)
        modified = true
    }
    fun descriptionProperty() = description
    
    // Debit
    fun getDebit() = debit.get()
    fun setDebit(value: BigDecimal?) {
        debit.set(value)
        if (value != null && value.compareTo(BigDecimal.ZERO) != 0) {
            credit.set(null) // Clear credit if debit is set
            originalEntry?.let {
                it.setDebit(true)
                it.setAmount(value)
            }
        }
        modified = true
    }
    fun debitProperty() = debit
    
    fun getDebitFormatted(): String {
        val d = debit.get()
        return if (d != null && d.compareTo(BigDecimal.ZERO) != 0) {
            CURRENCY_FORMAT.format(d)
        } else {
            ""
        }
    }
    
    // Credit
    fun getCredit() = credit.get()
    fun setCredit(value: BigDecimal?) {
        credit.set(value)
        if (value != null && value.compareTo(BigDecimal.ZERO) != 0) {
            debit.set(null) // Clear debit if credit is set
            originalEntry?.let {
                it.setDebit(false)
                it.setAmount(value)
            }
        }
        modified = true
    }
    fun creditProperty() = credit
    
    fun getCreditFormatted(): String {
        val c = credit.get()
        return if (c != null && c.compareTo(BigDecimal.ZERO) != 0) {
            CURRENCY_FORMAT.format(c)
        } else {
            ""
        }
    }
    
    // VAT
    fun getVatCode() = vatCode.get()
    fun setVatCode(value: String) = vatCode.set(value)
    fun vatCodeProperty() = vatCode
    
    fun updateVatFromAccount() {
        val acc = accountProperty.get()
        val rate = acc?.vatRate
        vatCode.set(
            if (rate != null && rate.compareTo(BigDecimal.ZERO) > 0) {
                "${rate.stripTrailingZeros().toPlainString()}%"
            } else {
                ""
            }
        )
    }
    
    // Entry access
    fun getEntry() = originalEntry
    fun setEntry(entry: Entry?) {
        this.originalEntry = entry ?: Entry()
    }
    
    fun isEmpty(): Boolean {
        return accountProperty.get() == null &&
               (description.get() == null || description.get().isEmpty()) &&
               debit.get() == null &&
               credit.get() == null
    }

    // Modified flag
    fun isModified() = modified
    fun setModified(value: Boolean) {
        modified = value
    }
    
    /**
     * Päivittää Entry-olion tiedot tämän rivin tiedoilla.
     */
    fun updateEntry() {
        val entry = originalEntry ?: Entry().also { originalEntry = it }
        
        entry.setRowNumber(rowNumber.get())
        entry.setDescription(description.get())
        
        accountProperty.get()?.let {
            entry.setAccountId(it.id)
        }
        
        val d = debit.get()
        val c = credit.get()
        
        when {
            d != null && d.compareTo(BigDecimal.ZERO) != 0 -> {
                entry.setDebit(true)
                entry.setAmount(d)
            }
            c != null && c.compareTo(BigDecimal.ZERO) != 0 -> {
                entry.setDebit(false)
                entry.setAmount(c)
            }
            else -> {
                entry.setAmount(BigDecimal.ZERO)
            }
        }
    }
    
    /**
     * Vaihda debet ja kredit keskenään.
     * Jos debetissä on arvo, siirretään se kreditiin ja päinvastoin.
     */
    fun toggleDebitCredit() {
        val d = debit.get()
        val c = credit.get()
        
        when {
            d != null && d.compareTo(BigDecimal.ZERO) != 0 -> {
                // Siirry debet → kredit
                credit.set(d)
                debit.set(null)
                originalEntry?.let {
                    it.setDebit(false)
                    it.setAmount(d)
                }
            }
            c != null && c.compareTo(BigDecimal.ZERO) != 0 -> {
                // Siirry kredit → debet
                debit.set(c)
                credit.set(null)
                originalEntry?.let {
                    it.setDebit(true)
                    it.setAmount(c)
                }
            }
            else -> {
                // Ei arvoa, ei vaihdeta
            }
        }
        modified = true
    }
    
    /**
     * Tarkista onko vienti debet-puolella.
     */
    fun isDebit(): Boolean {
        val d = debit.get()
        return d != null && d.compareTo(BigDecimal.ZERO) != 0
    }
}
